pkg/pushsync/pushsync.go

```diff
@@ -126,19 +130,35 @@ func (ps *PushSync) handler(ctx context.Context, p p2p.Peer, stream p2p.Stream)
 		return swarm.ErrInvalidChunk
 	}
 
-	span, _, ctx := ps.tracer.StartSpanFromContext(ctx, "pushsync-handler", ps.logger, opentracing.Tag{Key: "address", Value: chunk.Address().String()})
-	defer span.Finish()
-
-	// Get price we charge for upstream peer read at headler
+	// Get price we charge for upstream peer read at headler.
 	responseHeaders := stream.ResponseHeaders()
 	price, err := headerutils.ParsePriceHeader(responseHeaders)
-
+	// if not found in returned header, compute the price we charge for this chunk.
 	if err != nil {
-		// if not found in returned header, compute the price we charge for this chunk and
-		ps.logger.Warningf("push sync: peer %v no price in previously issued response headers: %v", p.Address, err)
+		ps.logger.Warningf("pushsync: peer %v no price in previously issued response headers: %v", p.Address, err)
 		price = ps.pricer.PriceForPeer(p.Address, chunk.Address())
 	}
 
+	// if the peer is closer to the chunk, we were selected for replication. Return early.
```

Comment:
If the chunk does not fall within our neighborhood depth, this is probably not the right conclusion to make.
can we merge this into the iswithindepth block below?

Reply:
Could the following be an issue:

The node replicating the chunk and the storer node have different ideas about the neighborhood, and the replicating node returns FALSE for IsWithinDepth. In this case, the node may forward it, instead of storing it. This may cause an infinite loop of some sorts.

Reply:
@metacertain

Reply:
@metacertain Esad is right and we cannot forward any chunk that came from a storer node which is farther away from us. The `IsWithinDepth` might miss that and lead to loops as Esad mentions. Let's leave this as is.

Reply:
why would the replicating node even store the chunk if it's not in it's neighborhood?

In general we dont forward to upstream if i understand correctly, but if there was a special node that would (for some reason) it would be able to spam our nodes with out-of-neighborhood content

Reply:
so, if the chunk is not within our ND, we will definietly charge a price for forwarding it.
so, when the receipt issuer pushes towards all peers in neighborhood, those who have a different idea about neighborhood depth will return a price > 0 in headers, and the receipt issuer skips those peers, doesn't push towards them.

Reply:
on the other hand, if the chunk does not fall into our neighborhood depth, that also means we have connections to forward the chunk towards, and I think that's fine, so I would still merge this block into the IsWithinDepth block

Reply:
We can add a IsWithinDepth check inside this distance compare if statement to guarantee that it is the neighborhood. BUT, putting it inside the IsWithinDepth is not good and will cause the loop problem we mentioned above.

Reply:
the way I see it, there could be a loop, if there was a node that does not behave well and forwards upstream for some reason.
 but even in this case the injected forked node, that forwards to upstream, would have to pay for the forwarding each time so that's okay.
 
 on the other hand, if we don't forward to neighborhood such a strange request, we just act rougishly and keep the fee without having managed to get the chunk to it's destination, no?

Reply:
@metacertain if we merge this block with the next, we might get a case where a node which is farther away from the chunk sent us a chunk and we still forward it and this is undesirable. Your solution of merging both blocks just institutionalizes this, which I believe is incorrect.

Reply:
I understand, but, this shouldn't happen with normal nodes that don't forward to other nodes farther away. (outside of the neighborhood as mentioned before)
In case against the odds it does happen, we should forward the chunk towards it's destination anyway, because we are not the destination in this case (if the chunk does not belong to our neighborhood). 
Or what should we do? We can refuse the forwarding alternatively, reset the stream, and by that, we would systematically prohibit this 'forked node' to push through nodes further away.

Reply:
okay maybe you are right it would be better to reset the stream in strange cases for now

Reply:
> this shouldn't happen with normal nodes that don't forward to other nodes farther away

so your assumption will not hold for malicious nodes. If I start a node that deliberately does upstream requests in order to generate loops, what then?

Reply:
it will be costly for it, and if the loop persists it will cause the request to timeout and probably be retried on a different route.
Not saying this is not mildly disruptive (though its a bit unlikely), and yes, that's why I said you are probably right, it's better not to accomodate for this kind of behaviour

so let's check distancecmp, if the requester is closer check if its in neighborhood depth (and do the currrent store+debit) 
if its not in neighborhood depth please introduce a stream reset instead, we don't want to debit or store in that case

Reply:
@metacertain I added the extra neighborhood check after a discussion with @acud . Thanks for the feedback!